src/org/jointheleague/level4/minesweeper/MineSweeper.java

```diff
@@ -0,0 +1,132 @@
+package org.jointheleague.level4.minesweeper;
+
+import java.awt.BorderLayout;
+import java.awt.Color;
+import java.awt.GridLayout;
+import java.util.Optional;
+import java.util.Random;
+import java.util.stream.IntStream;
+
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.JMenu;
+import javax.swing.JMenuBar;
+import javax.swing.JMenuItem;
+import javax.swing.JPanel;
+import javax.swing.SwingUtilities;
+import javax.swing.WindowConstants;
+
+/**
+ * Implement the Mine Sweeper game by following the instructions in the
+ * comments. All instructions are marked with "TODO".
+ */
+public class MineSweeper {
+    private static final int WIDTH = 10;
+    private static final int HEIGHT = 15;
+    private static final int CELL_SIZE = 20;
+    private static final int NUM_MINES = 15;
+    
+    final JFrame frame = new JFrame();
+    private final JButton[][] buttons = new JButton[HEIGHT][WIDTH];
+    private final Random rng = new Random();
+    private Optional<boolean[][]> mines = Optional.empty();
+    private int numCellsRemaining;
```

Comment:
This is the number of cells that do not have a mine and has not yet been opened. The name is not clear and can be misunderstood to mean the number of cells not yet opened, including cells with bombs. It is not clear what this field is used for. All we know is that it is decremented (line 106).